modules/core/src/main/java/org/apache/ignite/internal/client/thin/ClientCacheEntryListenerHandler.java

```
@@ -0,0 +1,266 @@
+            ClientChannel ch,
+            ClientOperation op,
+            long rsrcId,
+            ByteBuffer payload,
+            Exception err
+    ) {
+        if (op == ClientOperation.QUERY_CONTINUOUS_EVENT_NOTIFICATION && rsrcId == this.rsrcId) {
+            if (err == null && payload != null) {
+                BinaryInputStream in = BinaryByteBufferInputStream.create(payload);
+
+                int cnt = in.readInt();
+
+                List<CacheEntryEvent<? extends K, ? extends V>> evts = new ArrayList<>(cnt);
+
+                for (int i = 0; i < cnt; i++) {
+                    K key = utils.readObject(in, keepBinary);
+                    V oldVal = utils.readObject(in, keepBinary);
+                    V val = utils.readObject(in, keepBinary);
+                    byte evtTypeByte = in.readByte();
+
+                    EventType evtType = eventType(evtTypeByte);
+
+                    if (evtType != null) // Skip unknown event types.
+                        evts.add(new CacheEntryEventImpl<>(jCacheAdapter, evtType, key, oldVal, val));
+                }
+
+                locLsnr.onUpdated(evts);
+            }
+        }
+    }
+
+    /** {@inheritDoc} */
+    @Override public synchronized void close() {
+        ch.removeChannelCloseListener(chCloseLsnr);
+
+        if (clientCh != null) {
+            clientCh.removeNotificationListener(this);
+
+            if (!clientCh.closed())
+                clientCh.service(ClientOperation.RESOURCE_CLOSE, ch -> ch.out().writeLong(rsrcId), null);
+        }
+    }
+
+    /**
+     * Client channel.
+     */
+    public ClientChannel clientChannel() {
+        return clientCh;
+    }
+
+    /** */
+    private EventType eventType(byte evtTypeByte) {
+        switch (evtTypeByte) {
+            case 0: return EventType.CREATED;
+            case 1: return EventType.UPDATED;
+            case 2: return EventType.REMOVED;
+            case 3: return EventType.EXPIRED;
+            default: return null;
```

Should we log an error instead? Other event types are not valid.

---

We don't have logger on the client-side

---

Then we should close the channel with a `Throwable cause` that explains the problem.
Invalid event type means some kind of protocol breakdown, we should not silently continue.